benchmark/benchwrapper/main.py

```diff
@@ -0,0 +1,157 @@
+        """
+        with self.database.snapshot() as snapshot:
+            results = snapshot.execute_sql(request.query)
+
+        for _ in results:
+            # do nothing with the data
+            continue
+
+        return spanner_messages.EmptyResponse()
+
+    def Insert(self, request, context):
+        """Insert represents operations like Go's Client.Apply, Java's
+        DatabaseClient.writeAtLeastOnce, Python's transaction.commit, and Node's
+        Transaction.Commit.
+
+        It will typically be used to insert many items.
+        """
+        with self.database.batch() as batch:
+            batch.insert(
+                table="Singers",
+                columns=("SingerId", "FirstName", "LastName"),
+                values=[(i.id, i.first_name, i.last_name) for i in request.singers],
+            )
+
+        return spanner_messages.EmptyResponse()
+
+    def Update(self, request, context):
+        """Update represents operations like Go's
+        ReadWriteTransaction.BatchUpdate, Java's TransactionRunner.run,
+        Python's Batch.update, and Node's Transaction.BatchUpdate.
+
+        It will typically be used to update many items.
+        """
+        self.database.run_in_transaction(self.update_singers, request.queries)
+
+        return spanner_messages.EmptyResponse()
+
+    def update_singers(self, transaction, stmts):
+        """Method to execute batch_update in a transaction."""
+        status, row_cts = transaction.batch_update(stmts)
+
+        print(status, row_cts)
+
+
+def get_opts():
+    """Parse command line arguments."""
+    parser = OptionParser()
+    parser.add_option("-p", "--port", help="Specify a port to run on")
+
+    opts, _ = parser.parse_args()
+
+    return opts
+
+
+def validate_opts(opts):
+    """Validate command line arguments."""
+    if opts.port is None:
+        raise ValueError("Please specify a valid port, e.g., -p 5000 or "
+                         "--port 5000.")
+
+
+def start_grpc_server(num_workers, port):
+    """Method to start the GRPC server."""
+    # Instantiate the GRPC server.
+    server = grpc.server(futures.ThreadPoolExecutor(max_workers=num_workers))
+
+    # Instantiate benchwrapper service.
+    spanner_benchwrapper_service = SpannerBenchWrapperService()
+
+    # Add benchwrapper servicer to server.
+    spanner_service.add_SpannerBenchWrapperServicer_to_server(
+        spanner_benchwrapper_service, server)
+
+    # Form the server address.
+    addr = "localhost:{0}".format(port)
+
+    # Add the port, and start the server.
+    server.add_insecure_port(addr)
+    server.start()
+    server.wait_for_termination()
+
+
+def serve():
+    """Driver method."""
+    opts = get_opts()
+
+    validate_opts(opts)
+
+    if "SPANNER_EMULATOR_HOST" not in os.environ:
+        raise ValueError("This benchmarking server only works when connected "
+                         "to an emulator. Please set SPANNER_EMULATOR_HOST.")
```

Comment:
I suppose this check should be the first in the function. `get_opts()` and `validate_opts()` are fast enough, but they are useless without the env var set.

Reply:
Makes sense, done.